File: android/src/main/kotlin/com/cwt/rw_ble/RwBlePlugin.kt
package com.cwt.rw_ble

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import com.dh.ble.DHBleSdk
import com.dh.ble.callback.HealthSyncCallback
import com.dh.ble.model.HealthDataSyncBean
import com.dh.ble.constants.CmdConstants
import com.dh.ble.model.DHHeartRateModeSetModel

class RwBlePlugin: FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

    private lateinit var channel: MethodChannel
    private lateinit var healthChannel: EventChannel
    private lateinit var healthSink: EventChannel.EventSink

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "rw_ble/methods")
        healthChannel = EventChannel(binding.binaryMessenger, "rw_ble/health")
        channel.setMethodCallHandler(this)
        healthChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        healthChannel.setStreamHandler(null)
    }


    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        events?.let { healthSink = it }
        subscribeHealthData()
    }

    override fun onCancel(arguments: Any?) {}

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when(call.method){
            "startScan" -> {
                DHBleSdk.startScan()
                result.success(null)
            }
            "stopScan" -> {
                DHBleSdk.stopScan()
                result.success(null)
            }
            "connect" -> {
                val mac = call.argument<String>("mac") ?: ""
                DHBleSdk.connect(mac)
                result.success(null)
            }
            "login" -> {
                val userId = call.argument<String>("userId") ?: ""
                DHBleSdk.login(userId)
                result.success(null)
            }
            "getBattery" -> {
                result.success(DHBleSdk.getBattery())
            }
            "startHealthTest" -> {
                val type = call.argument<String>("type")
                DHBleSdk.startHealthTest(type)
                result.success(null)
            }
            "stopHealthTest" -> {
                val type = call.argument<String>("type")
                DHBleSdk.stopHealthTest(type)
                result.success(null)
            }
            "syncHistory" -> {
                val stress = call.argument<Boolean>("stress") ?: false
                val bloodSugar = call.argument<Boolean>("bloodSugar") ?: false
                DHBleSdk.syncHistory(stress, bloodSugar, object : HealthSyncCallback {
                    override fun onData(data: HealthData) {
                        healthSink.success(mapOf("type" to "history", "data" to data))
                    }
                })
                result.success(null)
            }
            "setHeartRateInterval" -> {
                val interval = call.argument<Int>("interval") ?: 30
                setHeartRateInterval(interval)
                result.success(null)
            }
            "setBoMonitoring" -> {
                val isOpen = call.argument<Boolean>("isOpen") ?: true
                setBoMonitoring(isOpen)
                result.success(null)
            }
            "setHrvMonitoring" -> {
                val isOpen = call.argument<Boolean>("isOpen") ?: true
                setHrvMonitoring(isOpen)
                result.success(null)
            }
            "setStressMonitoring" -> {
                val isOpen = call.argument<Boolean>("isOpen") ?: true
                setStressMonitoring(isOpen)
                result.success(null)
            }
            "setBloodSugarMonitoring" -> {
                val isOpen = call.argument<Boolean>("isOpen") ?: true
                setBloodSugarMonitoring(isOpen)
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }


    private fun subscribeHealthData() {
        // Heart Rate
        DHBleSdk.subscribeData(CmdConstants.JL_HR_DATA_TRANSFER_KEY, object : HealthDataBroCallback {
            override fun onResult(data: HealthDataSyncBean?) {
                data?.hrPartData?.lastOrNull()?.let { hr ->
                    healthSink.success(mapOf("type" to "hr", "value" to hr.hr, "timestamp" to System.currentTimeMillis()))
                }
            }
            override fun onFail(errorCode: Int) {}
            override fun onSuccess() {}
        })

        // SpO2
        DHBleSdk.subscribeData(CmdConstants.JL_BO_DATA_TRANSFER_KEY, object : HealthDataBroCallback {
            override fun onResult(data: HealthDataSyncBean?) {
                data?.boPartData?.lastOrNull()?.let { bo ->
                    healthSink.success(mapOf("type" to "spo2", "value" to bo.bo, "timestamp" to System.currentTimeMillis()))
                }
            }
            override fun onFail(errorCode: Int) {}
            override fun onSuccess() {}
        })

        // HRV
        DHBleSdk.subscribeData(CmdConstants.JL_HRV_DATA_TRANSFER_KEY, object : HealthDataBroCallback {
            override fun onResult(data: HealthDataSyncBean?) {
                data?.hrvPartData?.lastOrNull()?.let { hrv ->
                    healthSink.success(mapOf("type" to "hrv", "value" to hrv.hrv, "timestamp" to System.currentTimeMillis()))
                }
            }
            override fun onFail(errorCode: Int) {}
            override fun onSuccess() {}
        })

        // Stress
        DHBleSdk.subscribeData(CmdConstants.JL_PRESSURE_DATA_TRANSFER_KEY, object : HealthDataBroCallback {
            override fun onResult(data: HealthDataSyncBean?) {
                data?.pressurePartData?.lastOrNull()?.let { stress ->
                    healthSink.success(mapOf("type" to "stress", "value" to stress.pressure, "timestamp" to System.currentTimeMillis()))
                }
            }
            override fun onFail(errorCode: Int) {}
            override fun onSuccess() {}
        })

        // Blood Sugar
        DHBleSdk.subscribeData(CmdConstants.JL_BLOODSUGAR_DATA_TRANSFER_KEY, object : HealthDataBroCallback {
            override fun onResult(data: HealthDataSyncBean?) {
                data?.bloodSugarPartData?.lastOrNull()?.let { sugar ->
                    healthSink.success(mapOf("type" to "bloodSugar", "value" to sugar.value, "timestamp" to System.currentTimeMillis()))
                }
            }
            override fun onFail(errorCode: Int) {}
            override fun onSuccess() {}
        })
    }


    // SET MONITORING INTERVALS / ON-OFF
    private fun setHeartRateInterval(minutes: Int) {
        val model = DHHeartRateModeSetModel()
        model.isOpen = true
        model.startHour = 0
        model.startMinute = 0
        model.endHour = 23
        model.endMinute = 59
        model.interval = minutes
        DHBleSdk.setHeartRateMode(model) // SDK method
    }

    private fun setBoMonitoring(isOpen: Boolean) {
        val model = DHBoModeSetModel()
        model.isOpen = isOpen
        model.startHour = 0
        model.startMinute = 0
        model.endHour = 23
        model.endMinute = 59
        model.interval = 60
        DHBleSdk.setBoMode(model)
    }

    private fun setHrvMonitoring(isOpen: Boolean) {
        val model = DHHrvModeSetModel()
        model.isOpen = isOpen
        model.startHour = 0
        model.startMinute = 0
        model.endHour = 23
        model.endMinute = 59
        model.interval = 60
        DHBleSdk.setHrvMode(model)
    }

    private fun setStressMonitoring(isOpen: Boolean) {
        val model = DHStressModeSetModel()
        model.isOpen = isOpen
        model.startHour = 0
        model.startMinute = 0
        model.endHour = 23
        model.endMinute = 59
        model.interval = 60
        DHBleSdk.setStressMode(model)
    }

    private fun setBloodSugarMonitoring(isOpen: Boolean) {
        val model = DHBloodSugarModeSetModel()
        model.isOpen = isOpen
        model.startHour = 0
        model.startMinute = 0
        model.endHour = 23
        model.endMinute = 59
        model.interval = 60
        DHBleSdk.setBloodSugarMode(model)
    }



}

File: ios/Classes/RwBlePlugin.swift
import Flutter
import UIKit
import DHBleSDK

public class SwiftRwBlePlugin: NSObject, FlutterPlugin {

    // Store scanned peripherals
    private var scannedDevices: [String: DHPeripheralModel] = [:]

    private var healthSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "rw_ble/methods", binaryMessenger: registrar.messenger())
        let eventScan = FlutterEventChannel(name: "rw_ble/scan", binaryMessenger: registrar.messenger())
        let eventHealth = FlutterEventChannel(name: "rw_ble/health", binaryMessenger: registrar.messenger())

        let instance = SwiftRwBlePlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventHealth.setStreamHandler(instance)
        eventScan.setStreamHandler(instance)
    }

    private var methodChannel: FlutterMethodCall?

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startScan":
            DHBleCentralManager.startScan()
            result(nil)

        case "stopScan":
            DHBleCentralManager.stopScan()
            result(nil)

        case "connect":
              if let args = call.arguments as? [String: Any],
                    let mac = args["mac"] as? String,
                    let peripheral = scannedDevices[mac] {
                     DHBleCentralManager.connectDevice(with: peripheral)
                     result(nil)
                 } else {
                     result(FlutterError(code: "404", message: "Device not found", details: nil))
                 }
        case "login":
            if let args = call.arguments as? [String: Any],
               let userId = args["userId"] as? String {
                DHBleCommand.getJLBindInfoLogin { code, data in
                    result(nil)
                }
            }

        case "getBattery":
            DHBleCommand.getBattery { code, data in
                if let battery = data as? DHBatteryInfoModel {
                    result(battery.battery)
                } else {
                    result(0)
                }
            }

        case "startHealthTest":
            if let args = call.arguments as? [String: Any],
               let type = args["type"] as? String {
                startTest(type)
            }
            result(nil)

        case "stopHealthTest":
            if let args = call.arguments as? [String: Any],
               let type = args["type"] as? String {
                stopTest(type)
            }
            result(nil)

        case "syncHistory":
            if let args = call.arguments as? [String: Any] {
                let stress = args["stress"] as? Bool ?? false
                let bloodSugar = args["bloodSugar"] as? Bool ?? false
                let muslimPraise = args["muslimPraise"] as? Bool ?? false

                DHBleCommand.startDataSyncing(
                    _: stress,
                    supportBloodSugar: bloodSugar,
                    supportMuslimPraise: muslimPraise
                ) { code, progress, data in
                    guard code == 0 else { return }

                    if let array = data as? [Any] {
                        for model in array {
                            var event: [String: Any] = [
                                "type": "history",
                                "timestamp": Int(Date().timeIntervalSince1970)
                            ]

                            switch model {
                            case let step as DHDailyStepModel:
                                event["subType"] = "step"
                                event["step"] = step.step
                                event["calorie"] = step.calorie
                                event["distance"] = step.distance
                                event["items"] = step.items
                            case let sleep as DHDailySleepModel:
                                event["subType"] = "sleep"
                                event["duration"] = sleep.duration
                                event["beginTime"] = sleep.beginTime
                                event["endTime"] = sleep.endTime
                                event["items"] = sleep.items
                            case let hr as DHDailyHrModel:
                                event["subType"] = "hr"
                                event["items"] = hr.items
                            case let bo as DHDailyBoModel:
                                event["subType"] = "spo2"
                                event["items"] = bo.items
                            case let hrv as DHDailyHrvModel:
                                event["subType"] = "hrv"
                                event["items"] = hrv.items
                            case let stressModel as DHDailyPressureModel:
                                event["subType"] = "stress"
                                event["items"] = stressModel.items
                            case let bs as DHDailyBloodSugarModel:
                                event["subType"] = "bloodSugar"
                                event["items"] = bs.items
                            case let muslim as DHDailyMuslimCountModel:
                                event["subType"] = "muslimPraise"
                                event["muslimCount"] = muslim.muslimcount
                                event["items"] = muslim.items
                            default:
                                continue
                            }

                            self.healthSink?(event)
                        }
                    }
                }
            }
            result(nil)   // <--- close the "syncHistory" case properly

        case "setHeartRateInterval":
            if let args = call.arguments as? [String: Any],
               let interval = args["interval"] as? Int {
                setHeartRateInterval(interval)
            }
            result(nil)

        case "setBoMonitoring":
            if let args = call.arguments as? [String: Any],
               let isOpen = args["isOpen"] as? Bool {
                setBoMonitoring(isOpen)
            }
            result(nil)

        case "setHrvMonitoring":
            if let args = call.arguments as? [String: Any],
               let isOpen = args["isOpen"] as? Bool {
                setHRVMonitoring(isOpen)
            }
            result(nil)

        case "setStressMonitoring":
            if let args = call.arguments as? [String: Any],
               let isOpen = args["isOpen"] as? Bool {
                setStressMonitoring(isOpen)
            }
            result(nil)

        case "setBloodSugarMonitoring":
            if let args = call.arguments as? [String: Any],
               let isOpen = args["isOpen"] as? Bool {
                setBloodSugarMonitoring(isOpen)
            }
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        } // <--- closes the switch
    }


private func startTest(_ type: String) {
    let key: Int
    switch type {
    case "hr": key = 1
    case "spo2": key = 2
    case "hrv": key = 3
    case "stress": key = 4
    case "bloodSugar": key = 5
    default: return
    }
    DHBleCommand.controlOpen(1, dataType: key) { _, _ in }
}

private func stopTest(_ type: String) {
    let key: Int
    switch type {
    case "hr": key = 1
    case "spo2": key = 2
    case "hrv": key = 3
    case "stress": key = 4
    case "bloodSugar": key = 5
    default: return
    }
    DHBleCommand.controlOpen(0, dataType: key) { _, _ in }
}

    // MARK: - Monitoring Settings
    private func setHeartRateInterval(_ minutes: Int) {
              let model = DHHeartRateModeSetModel()
              model.isOpen = true
              model.startHour = 0
              model.startMinute = 0
              model.endHour = 23
              model.endMinute = 59
              model.interval = minutes
              DHBleCommand.setHeartRateMode(model) { code, data in
                  if code == 0 { print("Heart rate interval set to \(minutes) min") }
              }
          }

          private func setBoMonitoring(_ isOpen: Bool) {
              let model = DHBoModeSetModel()
              model.isOpen = isOpen
              model.startHour = 0
              model.startMinute = 0
              model.endHour = 23
              model.endMinute = 59
              model.interval = 60 // fixed 1 hour
              DHBleCommand.setBoMode(model) { code, data in
                  if code == 0 { print("Blood oxygen monitoring set: \(isOpen)") }
              }
          }

          private func setHRVMonitoring(_ isOpen: Bool) {
              let model = DHHrvModeSetModel()
              model.isOpen = isOpen
              model.startHour = 0
              model.startMinute = 0
              model.endHour = 23
              model.endMinute = 59
              model.interval = 60
              DHBleCommand.setHrvMode(model) { code, data in
                  if code == 0 { print("HRV monitoring set: \(isOpen)") }
              }
          }

          private func setStressMonitoring(_ isOpen: Bool) {
              let model = DHStressModeSetModel()
              model.isOpen = isOpen
              model.startHour = 0
              model.startMinute = 0
              model.endHour = 23
              model.endMinute = 59
              model.interval = 60
              DHBleCommand.setStressMode(model) { code, data in
                  if code == 0 { print("Stress monitoring set: \(isOpen)") }
              }
          }

          private func setBloodSugarMonitoring(_ isOpen: Bool) {
              let model = DHBloodSugarModeSetModel()
              model.isOpen = isOpen
              model.startHour = 0
              model.startMinute = 0
              model.endHour = 23
              model.endMinute = 59
              model.interval = 60
              DHBleCommand.setBloodSugarMode(model) { code, data in
                  if code == 0 { print("Blood sugar monitoring set: \(isOpen)") }
              }
          }

}

// MARK: Event Channel StreamHandler
extension SwiftRwBlePlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.healthSink = events
        setupHealthCallbacks()
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.healthSink = nil
        return nil
    }

    func setupHealthCallbacks() {
        NotificationCenter.default.addObserver(self, selector: #selector(healthDataChanged(_:)),
                                               name: NSNotification.Name(BluetoothNotificationHealthRingMeasureValueChange), object: nil)
    }

    @objc
    func healthDataChanged(_ ntf: Notification) {
      guard let info = ntf.userInfo,
            let type = info["dataType"] as? Int,
            let value = info["dataValue"] as? Int else { return }

      var event: [String: Any] = [
          "timestamp": Int(Date().timeIntervalSince1970)
      ]

      switch type {
        case 1: event["type"] = "hr"
        case 2: event["type"] = "spo2"
        case 3: event["type"] = "hrv"
        case 4: event["type"] = "stress"
        case 5: event["type"] = "bloodSugar"
        default: return
      }

      event["value"] = value
      healthSink?(event)
    }
}
